internal/provider/resource_mackerel_role.go

```diff
@@ -0,0 +1,147 @@
+package provider
+
+import (
+	"context"
+
+	"github.com/hashicorp/terraform-plugin-framework/path"
+	"github.com/hashicorp/terraform-plugin-framework/resource"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/planmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/stringdefault"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/stringplanmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/schema/validator"
+	"github.com/mackerelio-labs/terraform-provider-mackerel/internal/mackerel"
+)
+
+var (
+	_ resource.Resource                = (*mackerelRoleResource)(nil)
+	_ resource.ResourceWithConfigure   = (*mackerelRoleResource)(nil)
+	_ resource.ResourceWithImportState = (*mackerelRoleResource)(nil)
+)
+
+func NewMackerelRoleResource() resource.Resource {
+	return &mackerelRoleResource{}
+}
+
+type mackerelRoleResource struct {
+	Client *mackerel.Client
+}
+
+func (r *mackerelRoleResource) Metadata(_ context.Context, req resource.MetadataRequest, resp *resource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_role"
+}
+
+func (r *mackerelRoleResource) Schema(_ context.Context, _ resource.SchemaRequest, resp *resource.SchemaResponse) {
+	resp.Schema = schema.Schema{
+		Description: "This resource allows creating and management of Roles.",
+		Attributes: map[string]schema.Attribute{
+			"id": schema.StringAttribute{
+				Computed: true,
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.UseStateForUnknown(), // immutable
+				},
+			},
+			"service": schema.StringAttribute{
+				Description: "The name of a Service.",
+				Required:    true,
+				Validators: []validator.String{
+					mackerel.ServiceNameValidator(),
+				},
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.RequiresReplace(), // force new
+				},
+			},
+			"name": schema.StringAttribute{
+				Description: "The name of a Role.",
+				Required:    true,
+				Validators: []validator.String{
+					mackerel.RoleNameValidator(),
+				},
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.RequiresReplace(),
+				},
+			},
+			"memo": schema.StringAttribute{
+				Description: "Notes related to this role.",
+				Optional:    true,
+				Computed:    true,
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.RequiresReplace(),
```

Comment:
I made the issue because it was going to be a detriment, not a laughing matter.

- https://github.com/mackerelio-labs/terraform-provider-mackerel/issues/230

This pull request is intended to migrate from an existing implementation to based on the new framework and does not need to cover the issue.